### ISSUE TITLE
Type guards are not working in this case

### ISSUE THREAD

**Issue by adamrogas** [Microsoft/TypeScript#2001]
```typescript

func1() :Date{
     return new Date();
}
func2() :Date{
     return new Date();
}

interface options{
     dateRange: Date|Array<Date>;
}

init(opt?:options)
{
     console.log(opt);
}

init({ dateRange:[func1(),func2()] });


```

I get an error in VS2013 that says that type of Date[] is not assignable to type Date[] and type Date is not assignable to type Date

**Comment by adamrogas** [comment]
This issue is actually a different issue, sorry for the false alarm

**Comment by adamrogas** [comment]
The issue that caused this was the file that had func1 and func2 had some additional methods defined on a partial date interface. 

which made the date object returning from the function incompatible with the full date interface.

While I would view this as technically a but I am not sure what your thoughts are on that.

By moving the interface to a referenced file it solved the issue.

**Comment by RyanCavanaugh** [comment]
Can you clarify? I'm trying something like this (which does not error):

```ts
interface Date1 extends Date { foo; }
interface Date2 extends Date { bar; }

function func1(): Date1 {
     return undefined;
}
function func2(): Date2{
     return undefined;
}

interface options{
     dateRange: Date|Array<Date>;
}

function init(opt?:options)
{
     console.log(opt);
}

init({ dateRange:[func1(),func2()] });
```

**Comment by adamrogas** [comment]
I was extending the actual date interface so I had 

```typescript

interface Date
{
       today():Date;
}

```

in the same file as the functions and I had the definition for interface "options" in a separate file d.ts file 

when the methods were called from the file with the interface inside of it were only returning the definition of date as defined in the same file and not the composite Date interface

**Comment by RyanCavanaugh** [comment]
It sounds like those separate files were external modules and thus were returning a completely different `Date` shape. When you define an interface in an external module, it never augments an interface from the global scope. Moving those definitions to a different (non-external module) file was the correct fix.
